File: packages/lucid/src/lucid/capabilities.py
from __future__ import annotations

from typing import Any

from .discovery import ModelTarget, resolve_model_class
from .runtime import build_model_definition


DEFAULT_CONTROL_TOPIC = "wm.control"
DEFAULT_STATUS_TOPIC = "wm.status"


def selected_model_name(model: ModelTarget) -> str:
    definition = build_model_definition(resolve_model_class(model))
    return definition.name


def manifest(model: ModelTarget) -> dict[str, Any]:
    definition = build_model_definition(resolve_model_class(model))
    return definition.to_manifest()


def output_bindings(model: ModelTarget) -> list[dict[str, Any]]:
    definition = build_model_definition(resolve_model_class(model))
    bindings: list[dict[str, Any]] = []
    for output in definition.outputs:
        if output.kind in {"video", "audio"}:
            bindings.append(
                {
                    "name": output.name,
                    "kind": output.kind,
                    "track_name": output.name,
                }
            )
        else:
            bindings.append(
                {
                    "name": output.name,
                    "kind": output.kind,
                    "topic": f"wm.output.{output.name}",
                }
            )
    return bindings


def capabilities(
    *,
    control_topic: str = DEFAULT_CONTROL_TOPIC,
    status_topic: str = DEFAULT_STATUS_TOPIC,
    model: ModelTarget,
) -> dict[str, Any]:
    return {
        "control_topic": control_topic,
        "status_topic": status_topic,
        "manifest": manifest(model),
        "output_bindings": output_bindings(model),
    }

File: apps/coordinator/src/capabilities.rs
use std::{fs, path::Path};

use serde_json::Value;

use crate::models::{Capabilities, OutputBinding, CONTROL_TOPIC, STATUS_TOPIC};

pub fn load_manifest_from_path(path: &Path) -> Result<Value, String> {
    let raw = fs::read_to_string(path)
        .map_err(|err| format!("failed reading manifest {}: {err}", path.display()))?;
    serde_json::from_str(&raw)
        .map_err(|err| format!("invalid manifest JSON {}: {err}", path.display()))
}

pub fn build_capabilities(manifest: &Value) -> Capabilities {
    let output_bindings = manifest
        .get("outputs")
        .and_then(Value::as_array)
        .map(|outputs| {
            outputs
                .iter()
                .filter_map(|output| {
                    let name = output.get("name")?.as_str()?.to_string();
                    let kind = output.get("kind")?.as_str()?.to_string();
                    let mut binding = OutputBinding {
                        name,
                        kind: kind.clone(),
                        track_name: None,
                        topic: None,
                    };
                    if kind == "video" || kind == "audio" {
                        binding.track_name = Some(binding.name.clone());
                    } else {
                        binding.topic = Some(format!("wm.output.{}", binding.name));
                    }
                    Some(binding)
                })
                .collect::<Vec<_>>()
        })
        .unwrap_or_default();

    Capabilities {
        control_topic: CONTROL_TOPIC.to_string(),
        status_topic: STATUS_TOPIC.to_string(),
        manifest: manifest.clone(),
        output_bindings,
    }
}

#[cfg(test)]
mod tests {
    use serde_json::{json, Value};

    use super::build_capabilities;

    #[test]
    fn capabilities_include_track_binding_for_video_outputs() {
        let capabilities = build_capabilities(&json!({
            "model": {"name": "helios"},
            "outputs": [
                {"name": "main_video", "kind": "video"}
            ]
        }));

        assert_eq!(
            capabilities.manifest["model"]["name"],
            Value::String("helios".to_string())
        );
        assert_eq!(capabilities.output_bindings.len(), 1);
        assert_eq!(capabilities.output_bindings[0].name, "main_video");
        assert_eq!(
            capabilities.output_bindings[0].track_name.as_deref(),
            Some("main_video")
        );
    }
}
